File: xchange-idexjaxrs/src/main/java/org/knowm/xchange/idex/IdexTradeService.java
package org.knowm.xchange.idex;


import org.apache.commons.codec.binary.Hex;
import org.knowm.xchange.Exchange;
import org.knowm.xchange.currency.Currency;
import org.knowm.xchange.currency.CurrencyPair;
import org.knowm.xchange.dto.Order;
import org.knowm.xchange.dto.Order.OrderType;
import org.knowm.xchange.dto.trade.*;
import org.knowm.xchange.dto.trade.LimitOrder.Builder;
import org.knowm.xchange.idex.IdexExchange.Companion.IdexCurrencyMeta;
import org.knowm.xchange.idexjaxrs.dto.*;
import org.knowm.xchange.idexjaxrs.service.*;
import org.knowm.xchange.service.trade.TradeService;
import org.knowm.xchange.service.trade.params.CancelOrderParams;
import org.knowm.xchange.service.trade.params.TradeHistoryParams;
import org.knowm.xchange.service.trade.params.orders.OpenOrdersParams;
import si.mazi.rescu.RestProxyFactory;

import javax.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.MathContext;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.Arrays.asList;
import static org.knowm.xchange.dto.Order.OrderType.ASK;
import static org.knowm.xchange.dto.Order.OrderType.BID;
import static org.knowm.xchange.idex.IdexExchange.Companion.getCurrencyPair;
import static org.knowm.xchange.idex.IdexExchange.Companion.safeParse;
import static org.knowm.xchange.idex.IdexSignature.generateSignature;
import static org.web3j.crypto.Sign.SignatureData;

public class IdexTradeService implements TradeService {
    private final IdexExchange idexExchange;

    public IdexTradeService(IdexExchange idexExchange) {

        this.idexExchange = idexExchange;
    }

    @Override
    public OpenOrders getOpenOrders() {
        ReturnOpenOrdersApi proxy = RestProxyFactory.createProxy(ReturnOpenOrdersApi.class, getIdexExchange().getExchangeSpecification().getSslUri());

        OpenOrders ret = null;
        try {

            ReturnOpenOrdersResponse openOrdersResponse = proxy.openOrders(new OpenOrdersReq().address(getIdexExchange().getExchangeSpecification().getApiKey()));

            ret = new OpenOrders(
                    openOrdersResponse.stream().map(responseInner -> {
                        CurrencyPair currencyPair;
                        OrderType orderType = responseInner.getType() == IdexBuySell.BUY ? BID : ASK;

                        {

                            String market = responseInner.getMarket();
                            currencyPair = getCurrencyPair(market);
                        }
                        return new Builder(orderType, currencyPair)
                                .limitPrice(safeParse(responseInner.getPrice()))
                                .originalAmount(safeParse(responseInner.getAmount()))
                                .id(responseInner.getOrderHash())
                                .build();
                    }).collect(Collectors.toList())
            );
        } catch (Exception e) {
            e.printStackTrace();
        }
        return ret;
    }

    @Override
    public OpenOrders getOpenOrders(OpenOrdersParams openOrdersParams) {
        ReturnOpenOrdersApi proxy = RestProxyFactory.createProxy(ReturnOpenOrdersApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
        return null;
    }


    @Override
    public boolean cancelOrder(String s) {
        CancelApi proxy = RestProxyFactory.createProxy(CancelApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
        return false;
    }

    @Override
    public boolean cancelOrder(CancelOrderParams cancelOrderParams) {
        CancelApi proxy = RestProxyFactory.createProxy(CancelApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
        return false;
    }

    @Override
    public UserTrades getTradeHistory(TradeHistoryParams tradeHistoryParams) {
        ReturnTradeHistoryApi proxy = RestProxyFactory.createProxy(ReturnTradeHistoryApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
        return null;
    }

    @Override
    public Collection<Order> getOrder(String... strings) {
        ReturnOpenOrdersApi proxy = RestProxyFactory.createProxy(ReturnOpenOrdersApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
        return null;
    }

    @Override
    public OpenOrdersParams createOpenOrdersParams() {


        return null;
    }

    @Override
    public void verifyOrder(LimitOrder limitOrder) {
        throw new UnsupportedOperationException("Idex API doesn't support verify order");

    }

    @Override
    public void verifyOrder(MarketOrder marketOrder) {
        throw new UnsupportedOperationException("Idex API doesn't support verify order");

    }

    @Override
    public TradeHistoryParams createTradeHistoryParams() {

        return null;
    }

    @Override
    public String placeMarketOrder(MarketOrder marketOrder) {
        throw new UnsupportedOperationException("Idex API doesn't support market order");
    }

    @Override
    public String placeStopOrder(StopOrder stopOrder) {

        throw new UnsupportedOperationException("Idex API doesn't support stop order");
    }

    /**
     * returns OrderHash so you can fetch it and cancel it...  but there is a OrderNumber that you can intercept if you need to.
     */
    @Override
    public String placeLimitOrder(LimitOrder placeOrder) {
        OrderType type = placeOrder.getType();
        Currency baseCurrency = placeOrder.getCurrencyPair().base;
        Currency counterCurrency = placeOrder.getCurrencyPair().counter;
        BigDecimal originalAmount = placeOrder.getOriginalAmount();
        BigDecimal limitPrice = placeOrder.getLimitPrice();
        OrderReq orderReq = createNormalizedLimitOrderReq(baseCurrency, counterCurrency, type, limitPrice, originalAmount, null, null,
                null);
        try {
            RestProxyFactory.createProxy(OrderApi.class, getIdexExchange().getExchangeSpecification().getSslUri()).order(orderReq).getOrderHash();
        } catch (Exception e) {

            e.printStackTrace();
        }
        return null;
    }

    @NotNull
    public final OrderReq createNormalizedLimitOrderReq( @NotNull Currency baseCurrency, @NotNull Currency counterCurrency, @NotNull OrderType type, @NotNull BigDecimal limitPrice, @NotNull BigDecimal originalAmount, String contractAddress,  BigInteger nonce,BigInteger expires) {
        nonce = nonce == null ? BigInteger.valueOf(getIdexExchange().getNonceFactory().createValue()) : nonce;
        contractAddress = contractAddress == null ? contractAddress().getAddress() : contractAddress;
        expires = expires == null ? BigInteger.valueOf(100000) : expires;
        getIdexExchange().getExchangeMetaData().getCurrencies().get(baseCurrency);
        getIdexExchange().getExchangeMetaData().getCurrencies().get(counterCurrency);
        List<Currency> listOfCurrencies = asList(baseCurrency,/*OMG*/ counterCurrency/*ETH*/);
        if (type == ASK) Collections.reverse(listOfCurrencies);

        IdexCurrencyMeta buy_currency = (IdexCurrencyMeta) getIdexExchange().getExchangeMetaData().getCurrencies().get(listOfCurrencies.get(0));
        IdexCurrencyMeta sell_currency = (IdexCurrencyMeta) getIdexExchange().getExchangeMetaData().getCurrencies().get(listOfCurrencies.get(1));
        BigDecimal divide = originalAmount.divide(limitPrice, MathContext.DECIMAL128);
        BigDecimal amount_buy = divide.multiply(new BigDecimal("1e" + buy_currency.getDecimals()), MathContext.DECIMAL128);
        BigDecimal amount_sell = originalAmount.multiply(new BigDecimal("1e" + sell_currency.getDecimals()), MathContext.DECIMAL128);
        String buyc = buy_currency.getAddress();
        String sellc = sell_currency.getAddress();
        List<List<String>> hash_data = asList(
                asList("contractAddress", contractAddress, "address"),
                asList("tokenBuy", buyc, "address"),
                asList("amountBuy", amount_buy.toString(), "uint256"),
                asList("tokenSell", sellc, "address"),
                asList("amountSell", amount_sell.toString(), "uint256"),
                asList("expires", "" + expires, "uint256"),
                asList("nonce", "" + nonce, "uint256"),
                asList("address", "" + getApiKey(), "address")
        );

        SignatureData sig = generateSignature(getIdexExchange().getExchangeSpecification().getSecretKey(), hash_data);
        byte v = sig.getV();
        byte[] r = sig.getR();
        byte[] s = sig.getS();
        OrderReq orderReq = new OrderReq()
                .address(getApiKey())
                .nonce(nonce)
                .tokenBuy(buyc)
                .amountBuy(amount_buy.toString())
                .tokenSell(sellc)
                .amountSell(amount_sell.toString())
                .expires(expires)
                .r("0x" + new String(Hex.encodeHex(r)))
                .s("0x" + new String(Hex.encodeHex(s)))
                .v(BigInteger.valueOf(v & 0xffl));
        return orderReq;
    }

      ReturnContractAddressResponse contractAddress() {
        try {
            ReturnContractAddressApi proxy = RestProxyFactory.createProxy(ReturnContractAddressApi.class, getIdexExchange().getExchangeSpecification().getSslUri());
            return proxy.contractAddress();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public Exchange getIdexExchange() {
        return idexExchange;
    }

    public String getApiKey() {
        return getIdexExchange().getExchangeSpecification().getApiKey();

    }

}

File: xchange-idex/src/main/java/org/knowm/xchange/idex/IdexTradeService.kt
package org.knowm.xchange.idex

import org.apache.commons.codec.binary.*
import org.knowm.xchange.currency.*
import org.knowm.xchange.currency.Currency
import org.knowm.xchange.dto.*
import org.knowm.xchange.dto.Order.OrderType.*
import org.knowm.xchange.dto.marketdata.*
import org.knowm.xchange.dto.trade.*
import org.knowm.xchange.idex.dto.*
import org.knowm.xchange.idex.service.*
import org.knowm.xchange.idex.util.*
import org.knowm.xchange.service.trade.*
import org.knowm.xchange.service.trade.params.*
import org.knowm.xchange.service.trade.params.orders.*
import org.knowm.xchange.utils.*
import org.web3j.crypto.*
import org.web3j.crypto.Hash.*
import java.math.*
import java.math.BigDecimal.*
import java.util.*
import kotlin.text.Charsets.UTF_8

class IdexTradeService(val idexExchange: IdexExchange) : TradeService, TradeApi() {

    init {

        if (IdexExchange.debugMe) {
            apiClient = ApiClient()
            IdexExchange.setupDebug(apiClient)
        }
    }

    override fun cancelOrder(orderParams: CancelOrderParams): Boolean = when (orderParams) {
        is IdexCancelOrderParams -> cancel(orderParams).success!! == 1
        else -> throw ApiException("cancel order requires " + IdexCancelOrderParams::class.java.canonicalName)
    }

    val apiKey inline get() = idexExchange.exchangeSpecification.apiKey

    val idexServerNonce get() = nextNonce(NextNonceReq().address(apiKey)).nonce


    override fun placeLimitOrder(placeOrder: LimitOrder): String {
        when (placeOrder) {
            is IdexLimitOrder -> {
                val limitOrder: IdexLimitOrder = placeOrder
                val apiKey = idexExchange.exchangeSpecification.apiKey
                order(OrderReq()
                              .address(apiKey)
                              .nonce(idexServerNonce)
                              .amountBuy(limitOrder.amountBuy)
                              .amountSell(limitOrder.amountSell)
                              .tokenBuy(limitOrder.tokenBuy)
                              .tokenSell(limitOrder.tokenSell)
                              .r(limitOrder.r)
                              .s(limitOrder.s)
                              .v(limitOrder.v)
                ).orderNumber
            }
            else -> {

                val type = placeOrder.type
                val baseCurrency = placeOrder.currencyPair.base
                val counterCurrency = placeOrder.currencyPair.counter
                val originalAmount = placeOrder.originalAmount
                val limitPrice = placeOrder.limitPrice


                val orderReq = createNormalizedOrderReq(baseCurrency, counterCurrency, type, limitPrice, originalAmount)
                order(orderReq).orderNumber
            }
        }
        TODO("This method requires " + IdexLimitOrder::class.java.canonicalName);
    }

    public fun createNormalizedOrderReq(baseCurrency: Currency?,
                                        counterCurrency: Currency?,
                                        type: Order.OrderType?, limitPrice: BigDecimal,
                                        originalAmount: BigDecimal): OrderReq {
        idexExchange.exchangeMetaData.currencies[baseCurrency]
        idexExchange.exchangeMetaData.currencies[counterCurrency]


        var c = listOf(baseCurrency,//OMG
                       counterCurrency//ETH
        )
        if (type == ASK) c = c.reversed();

        val metaAccumulo = idexExchange.exchangeMetaData.currencies[c[0]] as IdexCurrencyMeta
        val metaReducto = idexExchange.exchangeMetaData.currencies[c[1]] as IdexCurrencyMeta


        var acquire = (originalAmount / limitPrice).multiply(
                "1e+${metaAccumulo.decimals.toBigDecimal()}".toBigDecimal()).toBigInteger()
        var reduce = (originalAmount).multiply(
                "1e+${metaReducto.decimals.toBigDecimal()}".toBigDecimal()).toBigInteger()

        val orderReq = OrderReq()
                .address(apiKey)
                .nonce(idexServerNonce)
                .tokenBuy(metaAccumulo.address)
                .amountBuy(acquire.toString())
                .tokenSell(metaReducto.address)
                .amountSell(reduce.toString())
                .expires(Random().nextInt())

        return orderReq
    }


    override fun createOpenOrdersParams() = OpenOrdersParams { it is IdexLimitOrder }


    override fun createTradeHistoryParams(): TradeHistoryParams = IdexTradeHistoryReq()

    override fun placeStopOrder(stopOrder: StopOrder?): String =
            throw org.knowm.xchange.exceptions.NotAvailableFromExchangeException()


    override fun cancelOrder(orderId: String) =
            cancel(CancelReq().run { orderHash(orderId).address(apiKey).nonce(idexServerNonce) }).success == 1


    override fun placeMarketOrder(marketOrder: MarketOrder): String =
            when (marketOrder) {
                is IdexMarketOrder -> {
                    val trade = trade(TradeReq().apply {
                        val m: IdexMarketOrder = marketOrder;
                        orderHash(m.id)
                                .amount(m.originalAmount.toPlainString())
                                .address(apiKey)
                                .nonce(idexServerNonce)
                                .r(m.r)
                                .s(m.s)
                                .v(m.v)
                    }
                    )
                    if (trade.isEmpty()) throw ApiException(
                            "Idex orders are done as all-or-nothing bulk arrays.  this trade " +
                                    "did not return any confirmations, all failed.")
                    trade.first().orderHash

                }
                else -> throw ApiException(
                        "Idex Market Orders require " + IdexMarketOrder::class.java.canonicalName)
            }

    override fun getTradeHistory(params: TradeHistoryParams): UserTrades {
        if (params !is IdexTradeHistoryReq) {
            throw  ApiException("tradehistory requires " + IdexTradeHistoryReq::class.java.canonicalName)
        }

        val m = params.market.split("_")
        val currencyPair = CurrencyPair(m[0], m[1])
        val tradeHistory = MarketApi().tradeHistory(params)
        val map = tradeHistory.map {
            UserTrade(enumValueOf(it.type),
                      it.amount.toBigDecimalOrNull() ?: ZERO,
                      currencyPair,
                      it.price.toBigDecimalOrNull() ?: ZERO,
                      DateUtils.fromISO8601DateString(it.date),
                      it.uuid, it.orderHash, ZERO, currencyPair.base)
        }
        var ret: UserTrades? = null
        if (params.lastId != null)
            params.lastId.let { ret = UserTrades(map, it, params.sort) }
        return ret ?: UserTrades(map, params.sort)
    }

    override fun verifyOrder(
            limitOrder: LimitOrder?) = throw org.knowm.xchange.exceptions.NotAvailableFromExchangeException()


    override fun verifyOrder(
            marketOrder: MarketOrder?) = throw org.knowm.xchange.exceptions.NotAvailableFromExchangeException()

    override fun getOpenOrders(): OpenOrders = OpenOrders(openOrders(OpenOrdersReq().address(apiKey)).map {

        val c2 = it.params.run { listOf(buySymbol, sellSymbol) }
        LimitOrder(BID[it.type], it.amount.toBigDecimalOrNull() ?: ZERO,
                   CurrencyPair(c2[0], c2[1]), it.orderHash.toString(), Date(),
                   it.price.toBigDecimalOrNull() ?: ZERO)
    })

    override fun getOpenOrders(params: OpenOrdersParams): OpenOrders = OpenOrders(
            getOpenOrders().openOrders.filter(params::accept))

    override fun getOrder(vararg orderIds: String?): MutableCollection<Order> =
            getOpenOrders { it.id in orderIds }.openOrders.toMutableList()

    companion object {
        class IdexMarketOrder(val r: String, val s: String, val v: String, type: OrderType?,
                              originalAmount: BigDecimal?,
                              currencyPair: CurrencyPair?, id: String?, timestamp: Date?, averagePrice: BigDecimal?,
                              cumulativeAmount: BigDecimal?, fee: BigDecimal?, status: OrderStatus?) :
                MarketOrder(type, originalAmount, currencyPair, id, timestamp, averagePrice, cumulativeAmount, fee,
                            status)

        /**
        v - ...
        r - ...
        s - (v r and s are the values produced by your private key signature, see above for details)
         */
        public class IdexLimitOrder(type: OrderType?, originalAmount: BigDecimal?,
                                    currencyPair: CurrencyPair?, id: String?,
                                    timestamp: Date?, limitPrice: BigDecimal?,
                                    val r: String,
                                    val s: String,
                                    val v: String,
                                    val orderHash: String,
                                    /**amountBuy (uint256) - The amount of the token you will receive when the order is fully filled
                                     */
                                    val amountBuy: String,
                                    /**amountSell (uint256) - The amount of the token you will give up when the order is fully filled
                                     */
                                    val amountSell: String,
                                    /**tokenBuy (address string) - The address of the token you will receive as a result of the trade
                                     */
                                    val tokenBuy: String,
                                    /**tokenSell (address string) - The address of the token you will lose as a result of the trade
                                     */
                                    val tokenSell: String,
                                    /**expires (uint256) - DEPRECATED this property has no effect on your limit order but is still REQUIRED to submit a limit order as it is one of the parameters that is hashed. It must be a numeric type
                                     */
                                    val expires: Long = Random().nextLong()

        ) : LimitOrder(type, originalAmount, currencyPair, id, timestamp, limitPrice)

        class IdexTradeHistoryReq(val sort: Trades.TradeSortType? = Trades.TradeSortType.SortByTimestamp,
                                  val lastId: Long? = null) : TradeHistoryParams, TradeHistoryReq()

        /**Generate v, r, s values from payload*/
        fun generateSignature(
                apiSecret: String,
                data: Iterable<List<String>>
        ): Sign.SignatureData? {
            // pack parameters based on type
            var sig_str = ""
            data.forEach { d ->

                var payload = d[1]
                if (d[2] == "address")
                // remove 0x prefix and convert to bytes
                    payload = payload.toLowerCase().split("0x").last()
                else if (d[2] == "uint256")
                // encode, pad and convert to bytes
                    payload = Hex.encodeHexString(d[1].toBigInteger().toByteArray()).toLowerCase()
                sig_str += payload
            }
            // hash the packed string
            val rawhash = sha3String(sig_str)

            // salt the hashed packed string
            val salted = sha3String("\u0019Ethereum Signed Message:\n32$rawhash")

            apiSecret.split("0x").last()

            val payloadBytesCrypted = salted.toByteArray(UTF_8)
            val apiSecretBytes = apiSecret.toLowerCase().split("0x").last().toByteArray(UTF_8)
            return Sign.signMessage(payloadBytesCrypted, ECKeyPair.create(apiSecretBytes))
        }
    }
}
